theano/sandbox/cuda/acdc.py

```diff
@@ -0,0 +1,309 @@
+#
+# The following is based on acdc_wrapper.cu
+# from https://github.com/mdenil/acdc-torch
+#
+
+import os
+
+import theano
+from theano.sandbox.cuda import GpuOp
+from theano.sandbox.cuda.basic_ops import as_cuda_ndarray_variable, gpu_contiguous
+
+import cuda_ndarray
+
+class FastACDC(GpuOp):
+
+    def c_headers(self):
+        return ['<stdio.h>', '<cufft.h>']
+
+    def c_libraries(self):
+        return ['cufft']
+
+    def make_node(self, x, A, Ab, D, Db):
+        x_ = gpu_contiguous(as_cuda_ndarray_variable(x))
+        A_ = gpu_contiguous(as_cuda_ndarray_variable(A))
+        Ab_ = gpu_contiguous(as_cuda_ndarray_variable(Ab))
+        D_ = gpu_contiguous(as_cuda_ndarray_variable(D))
+        Db_ = gpu_contiguous(as_cuda_ndarray_variable(Db))
+        for t in [x_, A_, Ab_, D_, Db_]:
+            assert t.dtype == 'float32'
+        return theano.Apply(self, [x_, A_, Ab_, D_, Db_], [x_.type()])
+
+    def grad(self, inputs, cost_grad):
+        inp, A, Ab, D, Db = inputs
+        gradOutput, = cost_grad
+        gradOutput = gpu_contiguous(gradOutput)
+        return FastACDCGrad()(inp, gradOutput, A, Ab, D, Db)
+
+    def c_support_code_apply(self, node, nodename):
+        # Based on a similar mechanism in sandbox/cuda/blas.py
+        #
+        # REMEMBER TO RAISE c_code_cache_version when changing any of
+        # these files
+        files = ['acdc.cu']
+        codes = [open(os.path.join(os.path.split(__file__)[0], f)).read()
+                for f in files]
+        return reduce(str.__add__, codes)
```

Comment:
This use the old inferface. A better one that don't request such hack but request some refactoring would be ideal, but currently this is working fine. If you want to do the refactoring, tell me and I can guide you. We didn't update all code in Theano as this is working fine.

You should add the c_code_cache_version() method:
```
def c_code_cache_version(self):
    return (1,)
```
Otherwise, it will get recompiled for each process.